OpenRA.Mods.Common/Traits/Mobile.cs

```diff
@@ -214,8 +214,6 @@ public int Facing
 		{
 			if (FromCell == ToCell)
 				return new[] { Pair.New(FromCell, FromSubCell) };
-			if (CanEnterCell(ToCell))
-				return new[] { Pair.New(ToCell, ToSubCell) };
 
 			return new[] { Pair.New(FromCell, FromSubCell), Pair.New(ToCell, ToSubCell) };
```

Comment:
Shouldn't this be `return new[] { Pair.New(ToCell, ToSubCell) };` now if we want the unit behaviour to remain the same?

Reply:
the idea behind the old code was that units occupied both the cell they were leaving *and* the cell they were entering while moving. This avoids units stacking on top of eachother if the movement is stopped (e.g. EMP) before it has left the starting cell.

Reply:
It is obvious that that was the idea. I'm just pointing out that that is _not_ how units have been behaving all this time&mdash;probably never have&mdash;because `CanEnterCell(ToCell)` is always true. It only gets set after checking whether the next cell in the path is accessible and then the actor immediately lays claim to it so no other actor can block it anymore.

The idea was also probably a bad one to begin with. It has no effect on the case where units stop before it has left the starting cell, because such a window doesn't exist in the first place. The new `mobile.ToCell` is set in the same tick as the `MovePart` is queued (which is not interruptible and always the first childactivity when it is queued) so an EMPed unit will always end up in `ToCell` no matter when the movement was stopped. If it hasn't started moving yet, `ToCell` is equal to `FromCell`.

What this _will_ affect is movement of groups of units. Because units in a column will have to wait till the unit in front of it has completely finished moving to the next cell before starting its move, it means tight formations are no longer possible. There will be odd gaps in the formation and starting a movement will be slow and chaotic.

In any case, this doesn't seem like the kind of behavioural change you would want to do this short before a release as part of a simple bugfix that is not, on the face of it, really related. And since it is just half a line difference to keep the status quo behaviour (intentional or not) while still fixing the bug, it seems prudent to do so.

Reply:
Since we don't allow changing subcells while remaining in the same cell, `(FromCell == ToCell) => (FromSubCell == ToSubCell)` right? In that case this entire method can simplify down to `return new[] { Pair.New(ToCell, ToSubCell) };`.

Reply:
> 
> 
> Since we don't allow changing subcells while remaining in the same cell, `(FromCell == ToCell) => (FromSubCell == ToSubCell)` right? In that case this entire method can simplify down to `return new[] { Pair.New(ToCell, ToSubCell) };`.

Looks like it.p

Reply:
@tovl is correct with the findings. Triggering `SpeedMultiplier: 0` with a weapon (no, TS EMP does not do this) proves.

Reply:
This chain seems like something worth following up, but in the context of fixing the cache bug I find it odd to change `OccupiedCells` at all. The bug is in the cache, and that's where we should target the fix. Changing `OccupiedCells` to appease the cache bug just replaces with a set of unknown bugs from changing what `OccupiedCells` means.

You're going to need to convince me there are no possible bad scenarios from removing two out of the three possible return results here. Are we aware of all the possible downstream affects from changing the fundamental definition of what cells a mobile unit occupies to consuming code? It feels like asking for regressions.

Reply:
The other solution is to add the old code for  `CanEntereCell` and the chain all the way down to the Locomotor where we dont use the cache for this call.

Reply:
The follow up bug that I posted in the Issue in caused by that the `OccupiedCells` is calling the cache which triggers the breakpoint check that was added as the TESTCASE here. We are getting from the cache where it hasn't been updated with the new movementtype yet (because we are in the call to do so).  **This should not cause any real desync because any next call to the cache will trigger a reevaluation on the cell.** 

But imo it would be cleaner to fix this by removing the OccupiedCells call to the locomotorcache, either by removing the ` CanEntereCell` from `OccupiedCells` which I always have felt a bit odd to have here, Its a bit backwards "I occupy these cells -- can I really be here?" This check should already be done in `Mobile/Move`.

The other fix would be the one I mentioned in the previous comment.

Reply:
> Are we aware of all the possible downstream affects from changing the fundamental definition of what cells a mobile unit occupies to consuming code? It feels like asking for regressions.

The change in behaviour can be easily seen in the radar bin - units between cells used to be displayed in both, while now only show in one. This will also impact actor visibility (that don't override the default `Footprint` type) and likely more.

Reply:
That means that `CanEnterCell(ToCell)` is not always true, which strikes me as odd/unexpected. That might even be bogus behaviour. But I tend to agree that a PR made for prep is not the ideal place to fix this now.
As for fixing it, would the following restore the wanted behaviour?
```
return FromCell == ToCell ? new[] { Pair.New(ToCell, ToSubCell) } : { Pair.New(FromCell, FromSubCell), Pair.New(ToCell, ToSubCell) };
```

Reply:
> That means that CanEnterCell(ToCell) is not always true

Correct, this has always had the definition quirk of “I can’t enter if I’m already in it”, and things break when this is changed.

Reply:
The transition of `OccupiedCells` when moving between two cells looks like this on the previous release (20190314). Imo it was best to look at it before we added the locomotor cache in case of any regressions.

```
FromCell
ToCell
FromCell-ToCell
FromCell
```

It only spends a short time on `ToCell`  and the majority on `FromCell-ToCell` while the move happens. I.e. in only got one line of `ToCell` logged.

Reply:
Can you update the transition list to note where the values of fromcell and tocell are the same or different?

Reply:
The only time the fromcell and tocell are the same is when `FromCell` is printed.

This is how the code looks:

```csharp
		public Pair<CPos, SubCell>[] OccupiedCells()
		{
			if (FromCell == ToCell)
			{
				Debug.WriteLine("FromCell");
				return new[] { Pair.New(FromCell, FromSubCell) };
			}

			if (CanEnterCell(ToCell))
			{
				Debug.WriteLine("ToCell");
				return new[] { Pair.New(ToCell, ToSubCell) };
			}

			Debug.WriteLine("FromCell-ToCell");
			return new[] { Pair.New(FromCell, FromSubCell), Pair.New(ToCell, ToSubCell) };
		}
```

Complete log for a move, only one actor on the map and dev visibility (for enabling radar). 

```
...
FromCell
FromCell
FromCell
ToCell
FromCell-ToCell
FromCell-ToCell
FromCell-ToCell
FromCell-ToCell
FromCell-ToCell
FromCell-ToCell
FromCell-ToCell
FromCell-ToCell
FromCell-ToCell
FromCell-ToCell
FromCell
FromCell
FromCell
...
```

Reply:
>  Are we aware of all the possible downstream affects from changing the fundamental definition of what cells a mobile unit occupies to consuming code? It feels like asking for regressions.

&rarr; #17290